frontend/catalyst/jax_tracer.py

```diff
@@ -166,7 +167,98 @@ def retrace_with_result_types(jaxpr: ClosedJaxpr, target_types: List[ShapedArray
     return ClosedJaxpr(jaxpr2, consts)
 
 
+def _apply_result_type_conversion(
+    ctx: JaxTracingContext,
+    jaxpr: ClosedJaxpr,
+    consts: List[Any],
+    target_types: List[ShapedArray],
+    num_implicit_outputs: int,
+) -> Tuple[InputSignature, OutputSignature]:
+    """Apply type conversion to the results of a Jaxpr program. Return full information about the
+    modified Jaxpr program.
+
+    Args:
+        ctx: Jax tracing context object.
+        jaxpr: The Jaxpr program to apply the conversion to. As any Jaxpr program, it might contain
+               both implicit and explicit arguments and results.
+        consts: List of constant values we need to know to trace this program.
+        target_types: List of types we want to convert the outputs of the program to. The list must
+                      match the number of outputs, except maybe the very last output if it is Qreg.
+        num_implicit_outputs: Number of implicit outputs found in the Jaxpr program.
+
+    Returns:
+        InputSignature: new input signature of the function
+        OutputSignature: new output signature of the function
+    """
+    with_qreg = len(target_types) > 0 and isinstance(target_types[-1], AbstractQreg)
+    args = [AbstractQreg()] if with_qreg else []
+
+    def _fun(*in_tracers):
+        out_tracers = eval_jaxpr(jaxpr, consts, *in_tracers)
+        out_tracers_, target_types_ = (
+            (out_tracers[:-1], target_types[:-1]) if with_qreg else (out_tracers, target_types)
+        )
+        out_promoted_tracers = [
+            (convert_element_type(tr, ty) if _abstractify(tr).dtype != ty else tr)
+            for tr, ty in zip(out_tracers_, target_types_)
+        ]
+        return out_promoted_tracers[num_implicit_outputs:] + (
+            [out_tracers[-1]] if with_qreg else []
+        )
+
+    _, in_sig, out_sig = trace_function(
+        ctx, _fun, *args, expansion_strategy=cond_expansion_strategy()
+    )
+
+    return in_sig, out_sig
+
+
+def _promote_jaxpr_types(types: List[List[Any]]) -> List[Any]:
+    # TODO: We seem to use AbstractQreg incorrectly, so JAX doesn't recognize it as a valid abstact
+    # value. One need to investigate how to use it correctly and remove the condition [1]. Should we
+    # add AbstractQreg into the `_weak_types` list of JAX?
```

Comment:
Can you comment a bit more on this?

Reply:
Our custom AbstractQreg happened to be incompatible with `jnp.promote_types`, while other built-in abstract values. I suspect we failed to match some expectation of Jax.

Reply:
I see. Thanks! Can you add this to the comment? I think it is important context that what is failing is the type promotion. I suspect that JAX was not made to handle non-numeric types and as such promotion is not even possible. But maybe, as you say it can be fixed with just adding either weak type or strict type. If anything, I think strict type might make more sense.

Reply:
Ok, I have updated the comment